src/ImageSharp/Formats/Pbm/BinaryEncoder.cs

```diff
@@ -0,0 +1,203 @@
+            int bytesPerPixel = 3;
+            MemoryAllocator allocator = configuration.MemoryAllocator;
+            using IMemoryOwner<byte> row = allocator.Allocate<byte>(width * bytesPerPixel);
+            Span<byte> rowSpan = row.GetSpan();
+
+            for (int y = 0; y < height; y++)
+            {
+                Span<TPixel> pixelSpan = image.GetPixelRowSpan(y);
+
+                PixelOperations<TPixel>.Instance.ToRgb24Bytes(
+                    configuration,
+                    pixelSpan,
+                    rowSpan,
+                    width);
+
+                stream.Write(rowSpan);
+            }
+        }
+
+        private static void WriteWideRgb<TPixel>(Configuration configuration, Stream stream, ImageFrame<TPixel> image)
+            where TPixel : unmanaged, IPixel<TPixel>
+        {
+            int width = image.Size().Width;
+            int height = image.Size().Height;
+            int bytesPerPixel = 6;
+            MemoryAllocator allocator = configuration.MemoryAllocator;
+            using IMemoryOwner<byte> row = allocator.Allocate<byte>(width * bytesPerPixel);
+            Span<byte> rowSpan = row.GetSpan();
+
+            for (int y = 0; y < height; y++)
+            {
+                Span<TPixel> pixelSpan = image.GetPixelRowSpan(y);
+
+                PixelOperations<TPixel>.Instance.ToRgb48Bytes(
+                    configuration,
+                    pixelSpan,
+                    rowSpan,
+                    width);
+
+                stream.Write(rowSpan);
+            }
+        }
+
+        private static void WriteBlackAndWhite<TPixel>(Configuration configuration, Stream stream, ImageFrame<TPixel> image)
+            where TPixel : unmanaged, IPixel<TPixel>
+        {
+            int width = image.Size().Width;
+            int height = image.Size().Height;
+            MemoryAllocator allocator = configuration.MemoryAllocator;
+            using IMemoryOwner<L8> row = allocator.Allocate<L8>(width);
+            Span<L8> rowSpan = row.GetSpan();
+
+            int previousValue = 0;
+            int startBit = 0;
+            for (int y = 0; y < height; y++)
+            {
+                Span<TPixel> pixelSpan = image.GetPixelRowSpan(y);
+
+                PixelOperations<TPixel>.Instance.ToL8(
+                    configuration,
+                    pixelSpan,
+                    rowSpan);
+
+                for (int x = 0; x < width;)
+                {
+                    int value = previousValue;
+                    for (int i = startBit; i < 8; i++)
+                    {
+                        if (rowSpan[x].PackedValue < 128)
+                        {
+                            value |= 0x80 >> i;
+                        }
+
+                        x++;
+                        if (x == width)
+                        {
+                            previousValue = value;
+                            startBit = (i + 1) % 8;
```

Comment:
Unsigned is not an option, as .NET did not provide bitshifting operators for unsigned.

Reply:
Unsigned is not an option, as .NET did not provide bitshifting operators for unsigned.

Reply:
You meant bit rotation for unsigned? Bit shifting is available for unsigned numbers.

Reply:
Line 182: `value |= 0x80 >> i`
Doesn't compile if `i` is unsigned

Reply:
Oh I see, I thought you were talking about something like  `value <<= i` for unsigned `value`.

gjoidl meant that JIT automatically can do `& 7` bithack if value is unsigned but if you`re sure that your `int` number is always positive (which it is as `i` is between `startbit` and 8) - you can manually apply this bithack via `& 7` instead of `% 8`.

Reply:
Indeed @br3aker , so I changed line 189 into:
`startBit = (i + 1) & 7; // Round off to below 8.`